File: src/engine/euConvergence.ts
/**
 * Moteur 3 (TypeScript) : Vecteur de Convergence Européenne & Sanctions
 */

export interface EUSanction {
  level: 'AVERTISSEMENT' | 'GEL_FONDS' | 'RUPTURE';
  code: string;
  title: string;
  withheldFundsMrd: number;
  spreadPenaltyBps: number;
  description: string;
}

export class EUConvergenceEngine {
  public euGravityCenter: number[] = [0.80, 0.85, 0.75, 0.70, 0.90];
  public weights: number[] = [0.30, 0.20, 0.15, 0.10, 0.25];
  public franceVector: number[] = [0.25, 0.60, 0.70, 0.65, 0.85];

  public applyPolicyDelta(delta: Partial<{
    disciplineBudgetaire: number;
    marcheUnique: number;
    climat: number;
    pesc: number;
    etatDeDroit: number;
  }>) {
    const raw = [
      delta.disciplineBudgetaire || 0,
      delta.marcheUnique || 0,
      delta.climat || 0,
      delta.pesc || 0,
      delta.etatDeDroit || 0
    ];
    this.franceVector = this.franceVector.map((val, idx) => {
      return Math.max(-1.0, Math.min(1.0, val + raw[idx]));
    });
  }

  public calculateEuDistance(): number {
    const weightedDiff = this.franceVector.map((val, idx) => {
      return this.weights[idx] * Math.pow(val - this.euGravityCenter[idx], 2);
    });
    return Math.sqrt(weightedDiff.reduce((a, b) => a + b, 0));
  }

  public evaluateSanctions(): { distance: number; sanctions: EUSanction[]; compliancePct: number } {
    const distance = this.calculateEuDistance();
    const sanctions: EUSanction[] = [];

    if (distance >= 0.45) {
      sanctions.push({
        level: 'AVERTISSEMENT',
        code: 'ART_258_AVIS_MOTIVE',
        title: 'Mise en demeure (Art. 258 TFUE)',
        withheldFundsMrd: 0,
        spreadPenaltyBps: 8.0,
        description: 'Bruxelles exige un plan de redressement budgétaire sous 2 mois.'
      });
    }
    if (distance >= 0.70) {
      sanctions.push({
        level: 'GEL_FONDS',
        code: 'GEL_NEXTGEN_EU',
        title: 'Gel des subventions européennes (12.4 Mrd€)',
        withheldFundsMrd: 12.4,
        spreadPenaltyBps: 25.0,
        description: 'Suspension des fonds NextGenEU et surveillance négative des agences de notation.'
      });
    }
    if (distance >= 0.95) {
      sanctions.push({
        level: 'RUPTURE',
        code: 'CJUE_ART_7',
        title: 'Saisine de la CJUE & Procédure Art. 7',
        withheldFundsMrd: 28.0,
        spreadPenaltyBps: 65.0,
        description: 'Crise constitutionnelle avec Bruxelles et fuite de capitaux.'
      });
    }

    const compliancePct = Math.max(0, Math.min(100, Math.round((1.0 - distance / 1.2) * 100)));
    return {
      distance: Number(distance.toFixed(3)),
      sanctions,
      compliancePct
    };
  }
}

File: engine_py/eu_convergence.py
"""
Module 3: Vecteur de Convergence Européenne et Système de Sanctions Automatiques
Version pure Python (Zéro dépendance)
"""

from dataclasses import dataclass
from typing import List, Dict, Tuple
import math
from math_utils import clip

@dataclass
class EUSanctionEvent:
    level: str
    code: str
    title: str
    withheld_funds_mrd: float
    capital_flight_factor: float
    spread_penalty_bps: float
    legal_deadline_weeks: int
    description: str

class EUConvergenceEngine:
    DIMENSION_NAMES = [
        "discipline_budgetaire",
        "marche_unique_aides_etat",
        "transition_ecologique",
        "politique_etrangere_pesc",
        "etat_de_droit_cjue"
    ]

    def __init__(self):
        self.eu_gravity_center = [0.80, 0.85, 0.75, 0.70, 0.90]
        self.metric_weights = [0.30, 0.20, 0.15, 0.10, 0.25]
        self.france_vector = [0.20, 0.60, 0.70, 0.65, 0.85]
        
        self.warning_threshold = 0.45
        self.infringement_threshold = 0.70
        self.sanction_threshold = 0.95

    def apply_policy_action(self, policy_vector_delta: Dict[str, float]) -> List[float]:
        delta = [
            policy_vector_delta.get("discipline_budgetaire", 0.0),
            policy_vector_delta.get("marche_unique_aides_etat", 0.0),
            policy_vector_delta.get("transition_ecologique", 0.0),
            policy_vector_delta.get("politique_etrangere_pesc", 0.0),
            policy_vector_delta.get("etat_de_droit_cjue", 0.0)
        ]
        self.france_vector = [clip(self.france_vector[i] + delta[i], -1.0, 1.0) for i in range(5)]
        return self.france_vector

    def calculate_eu_distance(self) -> float:
        weighted_sq_diff = [
            self.metric_weights[i] * ((self.france_vector[i] - self.eu_gravity_center[i]) ** 2)
            for i in range(5)
        ]
        return math.sqrt(sum(weighted_sq_diff))

    def evaluate_compliance_and_sanctions(self) -> Tuple[float, List[EUSanctionEvent]]:
        distance = self.calculate_eu_distance()
        sanctions: List[EUSanctionEvent] = []

        if distance >= self.warning_threshold:
            sanctions.append(EUSanctionEvent(
                level="AVERTISSEMENT",
                code="ART_258_AVIS_MOTIVE",
                title="Avis motivé de la Commission Européenne : Mise en demeure",
                withheld_funds_mrd=0.0,
                capital_flight_factor=0.02,
                spread_penalty_bps=8.0,
                legal_deadline_weeks=8,
                description="Bruxelles exige un mémoire en réponse et des engagements correctifs sous 2 mois."
            ))

        if distance >= self.infringement_threshold:
            sanctions.append(EUSanctionEvent(
                level="SANCTION_FINANCIERE",
                code="GEL_FONDS_NEXTGEN_EU",
                title="Gel immédiat des subventions NextGenerationEU et FEDER",
                withheld_funds_mrd=12.4,
                capital_flight_factor=0.08,
                spread_penalty_bps=25.0,
                legal_deadline_weeks=4,
                description="Suspension des décaissements européens. Les agences de notation placent la France sous surveillance négative."
            ))

        if distance >= self.sanction_threshold:
            sanctions.append(EUSanctionEvent(
                level="RUPTURE_INSTITUTIONNELLE",
                code="CJUE_ASTREINTES_ART7",
                title="Saisine de la CJUE : Astreintes journalières record et alerte Article 7",
                withheld_funds_mrd=28.0,
                capital_flight_factor=0.22,
                spread_penalty_bps=65.0,
                legal_deadline_weeks=2,
                description="Crise existentielle avec l'Union. Fuite massive de capitaux vers les bunds allemands et le Trésor américain."
            ))

        return distance, sanctions
